### DISCUSSION TITLE
What an addon can do, and what it can't

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#1641] [Discussion]
I suppose that the nature of an addon (working on the output of the Scratch source code) makes it easy to do something but it makes (maybe?) impossible to do something else. I never worked on a browser extensions but, if I understand it correctly, you are not able to create a new block by means of an addon, are you? Or I am wrong and you can fully overwrite the Scratch source code? 

I think that Scratch 1.4 has been one of the best Scratch versions to be used by a teacher. Having tools such as the enlarge/shrink tools, the possibility to rotate or change the dimension of a sprite by rightclicking it or by dragging the blue sprite direction line on the sprite thumbnail, the possibility to copy areas of the Stage to make them a new sprite or a new costume, these possibilities, and others, make teaching Scratch much easier than by using a "pure" Scratch 3.0. Then Scratch Addons added a lot of very useful options.
 
Are there other things that cannot be obviously done by means of a Scratch addon? I would love hearing from the Scratch addons developers about things that they would like to add to Scratch but cannot be added by using the Chrome extension mechanism.

Thanks in advance

**Comment by WorldLanguages** [comment]
Pretty much anything can be added, but there's a line where it becomes more convenient to fork the editor itself, instead of running our own JS on top of the official editor. For those cases, check out TurboWarp.
We can probably create new blocks, but of course users without Scratch Addons wouldn't be able to use the block properly, so it's pretty useless.

**Reply by sfederici** [reply]
Thanks for your reply!

> We can probably create new blocks, but of course users without Scratch Addons wouldn't be able to use the block properly, so it's pretty useless.

This is not always the case. Indeed I'm thinking to new blocks that can be easily implemented by other Scratch blocks (so that they can be saved in the sb3 file as the original blocks, maybe with a comment that will allow the editor to show the new blocks if the project is open by a Scratch Addon user). For instance, to newbies that are learning Scratch but don't think they are very interested in computer programming the IF block inside the FOREVER block is something that they learn by rote memory, but do not really understand. Instead, the old FOREVER IF block (lost since Scratch 2.0) to them is immediately clear and they will use it very easily. Then, when they become more experts, they can see that replacing this block with two blocks can be convenient if they want to avoid to create too many very short scripts in their projects.

**Reply by cobaltt7** [reply]
@sfederici for that, see #817
